llvm/lib/Transforms/Utils/LoopPeel.cpp

```diff
@@ -351,11 +351,19 @@ static unsigned countToEliminateCompares(Loop &L, unsigned MaxPeelCount,
     MaxPeelCount =
         std::min((unsigned)SC->getAPInt().getLimitedValue() - 1, MaxPeelCount);
 
-  auto ComputePeelCount = [&](Value *Condition) -> void {
+  std::function<void(Value *)> ComputePeelCount =
+      [&](Value *Condition) -> void {
     if (!Condition->getType()->isIntegerTy())
       return;
 
     Value *LeftVal, *RightVal;
+    if (match(Condition, m_And(m_Value(LeftVal), m_Value(RightVal))) ||
+        match(Condition, m_Or(m_Value(LeftVal), m_Value(RightVal)))) {
+      ComputePeelCount(LeftVal);
+      ComputePeelCount(RightVal);
```

Comment:
It would be safer to use a worklist instead of recursion here.

Reply:
What I had in mind here is stack overflow due to deep recursion. But yes, having a visited set is also a good idea, as it ensures there is no exponential blowup for some degenerate IR.

Reply:
I added a depth limit, and some new test cases.